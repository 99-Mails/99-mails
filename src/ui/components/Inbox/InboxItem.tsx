import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Flex,
  Card,
  Link,
  Text,
  CardBody,
  CardHeader,
  CardFooter,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
  AccordionItemProps,
  forwardRef,
} from "@chakra-ui/react";
import type { MailSummary } from "@/domains/Mail";
import { FancyDate } from "@/lib/date";
import { memo } from "react";

const InboxHeadItem = ({ text, value }: { text: string; value: string }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

type InboxItemProps = AccordionItemProps & {
  mail: MailSummary;
};

const InboxItem = memo(forwardRef<InboxItemProps, "div">((props, ref) => {
  const { mail, ...restProps } = props;
  const { headerSubject, toAddr, fromAddr, text, downloadUrl } = mail;

  return (
    <AccordionItem ref={ref} {...restProps} w="full" >
      {({ isExpanded }) => (
        <Card
          boxShadow="none"
          border="1px"
          borderColor="gray.200"
          borderRadius="0"
        >
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem
                      text="Received at:"
                      value={new FancyDate().toISOString()}
                    />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl.href} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
}));

export { InboxItem, InboxHeadItem };
