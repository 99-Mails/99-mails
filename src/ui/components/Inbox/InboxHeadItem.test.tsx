import { it, expect } from "vitest";
import { render } from "../../../tests/test-utils";
import { InboxHeadItem } from "./InboxItem";

it("should render correctly", () => {
  const { asFragment } = render(
    <InboxHeadItem text="subject" value="this is a mail" />
  );
  expect(asFragment()).toMatchSnapshot();
});
