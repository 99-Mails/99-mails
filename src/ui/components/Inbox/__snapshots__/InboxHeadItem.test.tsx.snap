// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1utx3w7"
  >
    <b
      class="chakra-text css-0"
    >
      subject
    </b>
    <p
      class="chakra-text css-0"
    >
      this is a mail
    </p>
  </div>
</DocumentFragment>
`;
